src/transformStateWithClones.js

```diff
@@ -6,8 +6,33 @@
  *
  * @return {Object[]}
  */
+
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateCopy = { ...state };
+  const history = [];
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        Object.assign(stateCopy, action.extraData);
+        break;
+      case 'removeProperties':
+        for (const removeElement of action.keysToRemove) {
+          delete stateCopy[removeElement];
+        }
+        break;
+      case 'clear':
+        for (const stateCopyKey in stateCopy) {
+          delete stateCopy[stateCopyKey];
+        }
```

Comment:
According to the checklist, you should not mutate an object or arrays. In the 'clear' case, instead of deleting all properties from the current stateCopy, you should create a new empty object. This is important to ensure that previous states in the history array are not affected by the clear action.